modules/geochemistry/src/outputs/EquilibriumReactionSolverOutput.C

```diff
@@ -0,0 +1,417 @@
+        getParam<bool>("ionic_str_using_basis_only")),
+    _gac(_is),
+    _egs(_mgd,
+         _gac,
+         _is,
+         _swapper,
+         getParam<std::vector<std::string>>("swap_out_of_basis"),
+         getParam<std::vector<std::string>>("swap_into_basis"),
+         getParam<std::string>("charge_balance_species"),
+         getParam<std::vector<std::string>>("constraint_species"),
+         getParam<std::vector<Real>>("constraint_value"),
+         getParam<MultiMooseEnum>("constraint_meaning"),
+         getParam<Real>("temperature"),
+         getParam<unsigned>("extra_iterations_to_make_consistent"),
+         getParam<Real>("min_initial_molality")),
+    _solver(_mgd,
+            _egs,
+            _is,
+            getParam<Real>("abs_tol"),
+            getParam<Real>("rel_tol"),
+            getParam<unsigned>("max_iter"),
+            getParam<Real>("max_initial_residual"),
+            getParam<Real>("swap_threshold") * getParam<Real>("abs_tol"),
+            getParam<std::vector<std::string>>("prevent_precipitation"),
+            getParam<Real>("max_ionic_strength"),
+            getParam<unsigned>("ramp_max_ionic_strength")),
+    _original_redox_lhs(_mgd.redox_lhs),
+    _precision(getParam<unsigned int>("precision")),
+    _stoi_tol(getParam<Real>("stoichiometry_tolerance")),
+    _verbose(getParam<bool>("verbose")),
+    _mol_cutoff(getParam<Real>("mol_cutoff")),
+    _temperature(getParam<Real>("temperature")),
+    _nernst_swap_out_of_basis(getParam<std::vector<std::string>>("nernst_swap_out_of_basis")),
+    _nernst_swap_into_basis(getParam<std::vector<std::string>>("nernst_swap_into_basis"))
+{
+  if (_nernst_swap_out_of_basis.size() != _nernst_swap_into_basis.size())
+    paramError("nernst_swap_out_of_basis", "must be of same size as nernst_swap_into_basis");
+}
+
+void
+EquilibriumReactionSolverOutput::output(const ExecFlagType & type)
+{
+  if (!shouldOutput(type))
+    return;
+
+  // solve the geochemical system
+  std::stringstream ss;
+  unsigned tot_iter;
+  Real abs_residual;
+  _solver.solveSystem(ss, tot_iter, abs_residual);
+  if (_verbose)
+    _console << ss.str();
+
+  // retrieve information
+  const std::vector<Real> basis_molality = _egs.getSolventMassAndFreeMolalityAndMineralMoles();
+  std::vector<Real> basis_activity;
+  for (unsigned i = 0; i < _num_basis; ++i)
+    basis_activity.push_back(_egs.getBasisActivity(i));
+  std::vector<Real> basis_act_coef;
+  for (unsigned i = 0; i < _num_basis; ++i)
+    basis_act_coef.push_back(_egs.getBasisActivityCoefficient(i));
+  const std::vector<Real> bulk_moles = _egs.getBulkMoles();
+  std::vector<Real> eqm_molality;
+  for (unsigned j = 0; j < _num_eqm; ++j)
+    eqm_molality.push_back(_egs.getEquilibriumMolality(j));
+  std::vector<Real> eqm_act_coef;
+  for (unsigned j = 0; j < _num_eqm; ++j)
+    eqm_act_coef.push_back(_egs.getEquilibriumActivityCoefficient(j));
+  const std::vector<Real> eqm_SI = _egs.getSaturationIndices();
```

Comment:
You should probably have getters for all of these values in the form of the entire vector, so that this simplifies to

```
const auto & basis_molality = _egs.getSolventMassAndFreeMolalityAndMineralMoles();
const auto & basis_activity = _egs.getBasisActivity();
const auto & basis_act_coef = _egs.getBasisActivityCoefficients();
...
```

Reply:
OK, will do

Reply:
Done